File: src/main/java/org/capturecoop/cclogger/CCLogger.java
package org.capturecoop.cclogger;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.capturecoop.ccutils.utils.CCStringUtils;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.time.LocalDateTime;
import java.util.ArrayList;

public class CCLogger {
    private static boolean enabled = false;
    private static boolean paused = false;
    private static File logFile;
    private static final int MAX_LEVEL_LENGTH = CCLogLevel.WARNING.toString().length();
    private static final ArrayList<String> preFileMessages = new ArrayList<>(); //These are printed but not written right away
    private static final ArrayList<CCLogMessage> pausedMessages = new ArrayList<>(); //These need to be constructed once ready
    private static String logFormat = "[%hour%:%minute%:%second%:%ms%] [%level%]%levelspace% [%filename%.%method%:%line%]: %message%";
    private static String htmlLog = "";
    private static String gitHubCodePathURL = null; //Example: https://github.com/CaptureCoop/SnipSniper/tree/<HASH HERE>/src/main/java/"
    private static String gitHubCodeClassPath; //Example: org.snipsniper -> If set only messages that contain this classpath get the @ link
    private static CCDebugConsole console;
    private static final String THIS_CLASSPATH = "org.capturecoop.cclogger.CCLogger";

    // Debug    -> Log everything + debug messages
    // Info     -> Log everything
    // Warning  -> Log warnings and errors
    // Error    -> Log errors
    private static CCLogLevel logLevel = CCLogLevel.INFO;

    private CCLogger() {}

    public static void log(String message) {
        log(message, CCLogLevel.INFO);
    }

    public static void log(String message, CCLogLevel level) {
        if(!enabled)
            return;

        if(paused) {
            pausedMessages.add(new CCLogMessage(level, message, LocalDateTime.now(), getStackTrace(), false));
            return;
        }

        logInternal(message, level, LocalDateTime.now(), getStackTrace());
    }

    public static void log(String message, CCLogLevel level, Object... args) {
        if(!enabled)
            return;

        log(org.capturecoop.ccutils.utils.CCStringUtils.format(message, args), level);
    }

    public static void logStacktrace(CCLogLevel level) {
        if(!enabled)
            return;

        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        final int STACKTRACE_START = 2;

        StringBuilder stackTraceString = new StringBuilder();
        for(int i = STACKTRACE_START; i < stackTrace.length; i++) {
            String trace = stackTrace[i].toString();
            if(trace.contains("net.snipsniper"))
                stackTraceString.append(trace).append("\n");
        }
        logStacktraceInternal(stackTraceString.toString(), level);
    }

    public static void logStacktrace(Throwable exception, CCLogLevel level) {
        if(!enabled)
            return;

        logStacktraceInternal(ExceptionUtils.getStackTrace(exception), level);
    }

    private static void logStacktraceInternal(String message, CCLogLevel level) {
        if(!enabled)
            return;

        if(paused) {
            pausedMessages.add(new CCLogMessage(level, message, LocalDateTime.now(), null, true));
            return;
        }

        System.out.println(message);
        writeToFile(message);
        htmlLog += "<p style='margin-top:0; white-space: nowrap;'><font color='" + getLevelColor(level) + "'>" + org.apache.commons.text.StringEscapeUtils.escapeHtml4(message).replaceAll("\n", "<br>") + "</font></p>";
        htmlLog += "<br>";
    }

    private static StackTraceElement getStackTrace() {
        //We loop through the stacktrace till we get to whoever called the logger. We skip the first element as its java.lang.Thread
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        int startIndex;
        for(startIndex = 1; startIndex < stackTrace.length; startIndex++) {
            if(stackTrace[startIndex].getClassName().startsWith(THIS_CLASSPATH))
                startIndex++;
            else
                break;
        }
        return stackTrace[startIndex];
    }

    //The reason for this is that this way we can take index 3 of stack trace at all times
    private static void logInternal(String message, CCLogLevel level, LocalDateTime time, StackTraceElement currentStackTrace) {
        if(!enabled)
            return;

        if(level == CCLogLevel.DEBUG && logLevel == CCLogLevel.DEBUG)
            return;

        StringBuilder msg = new StringBuilder(logFormat);
        msg = new StringBuilder(CCStringUtils.formatDateTimeString(msg.toString(), time));

        String levelString = level.toString();

        if(levelString.length() <= MAX_LEVEL_LENGTH) {
            msg = new StringBuilder(msg.toString().replace("%levelspace%", StringUtils.repeat(" ", MAX_LEVEL_LENGTH - levelString.length())));
        } else {
            levelString = levelString.substring(0, MAX_LEVEL_LENGTH);
            msg = new StringBuilder(msg.toString().replace("%levelspace%", ""));
        }

        String classFilename = currentStackTrace.getFileName();
        if(classFilename != null)
            classFilename = classFilename.replaceAll(".java" ,"");

        msg = new StringBuilder(msg.toString().replace("%filename%", classFilename));
        msg = new StringBuilder(msg.toString().replace("%method%", currentStackTrace.getMethodName()));
        msg = new StringBuilder(msg.toString().replace("%line%", currentStackTrace.getLineNumber() + ""));

        msg = new StringBuilder(msg.toString().replace("%levelspace%", ""));
        msg = new StringBuilder(msg.toString().replace("%level%", levelString));
        msg = new StringBuilder(msg.toString().replace("%message%", message));

        System.out.println(msg.toString().replaceAll("%newline%", "\n"));

        String finalMsg = org.apache.commons.text.StringEscapeUtils.escapeHtml4(msg.toString()).replaceAll(" ", "&nbsp;");
        finalMsg = finalMsg.replaceAll("%newline%", "<br>");
        if(gitHubCodePathURL != null && (gitHubCodeClassPath == null || gitHubCodeClassPath.isEmpty() || currentStackTrace.getClassName().contains(gitHubCodeClassPath))) {
            String link = gitHubCodePathURL + currentStackTrace.getClassName().replaceAll("\\.", "/") + ".java#L" + currentStackTrace.getLineNumber();
            finalMsg = finalMsg.replace(":" + currentStackTrace.getLineNumber() + "]", ":" + currentStackTrace.getLineNumber() + " <a href='" + link + "'>@</a>]");
        }
        String htmlLine = "<p style='margin-top:0; white-space: nowrap;'><font color='" + getLevelColor(level) + "'>" + finalMsg + "</font></p>";

        htmlLog += htmlLine;

        if(console != null)
            console.update();

        msg.append("\n");
        writeToFile(msg.toString());
    }

    public static void writeToFile(String message) {
        if(logFile != null) {
            try {
                Files.write(Paths.get(logFile.getAbsolutePath()), message.getBytes(), StandardOpenOption.APPEND);
            } catch (IOException ioException) {
                String path = logFile.getAbsolutePath();
                logFile = null;
                CCLogger.log("Could not write to logfile at \"%c\". Disabling logFile & Printing to console as well just in case!", CCLogLevel.ERROR, path);
                CCLogger.logStacktrace(ioException, CCLogLevel.ERROR);
            }
        } else {
            preFileMessages.add(message);
        }
    }

    public static String getLevelColor(CCLogLevel level) {
        String color = "white";
        if(level == CCLogLevel.WARNING)
            color = "yellow";
        else if(level == CCLogLevel.ERROR)
            color = "red";
        return color;
    }

    public static String getHTMLLog() {
        return htmlLog;
    }

    public static void setLogFile(File logFile) {
        CCLogger.logFile = logFile;
        if(!logFile.exists()) {
            try {
                if (logFile.createNewFile()) {
                    CCLogger.log("Set log file does not exist. Creating: %c", CCLogLevel.INFO, logFile.getAbsolutePath());
                }
            } catch (IOException ioException) {
                CCLogger.log("Set log file does not exist and could not be created. File: %c", CCLogLevel.ERROR, logFile.getAbsolutePath());
                CCLogger.logStacktrace(ioException, CCLogLevel.ERROR);
            }
        }
        refreshPreFileMessages();
    }

    public static void enableDebugConsole(boolean enabled) {
        if(enabled) {
            if(console == null)
                console = new CCDebugConsole();
            console.update();
        } else {
            console.setVisible(false);
            console.dispose();
            console = null;
        }
    }

    private static void refreshPreFileMessages() {
        if(!preFileMessages.isEmpty()) {
            for(String m : preFileMessages) {
                writeToFile(m);
            }
            preFileMessages.clear();
        }
    }

    public static void setPaused(boolean paused) {
        CCLogger.paused = paused;
        if(!paused) {
            for(CCLogMessage msg : pausedMessages) {
                if(!msg.isStacktrace()) {
                    logInternal(msg.getMessage(), msg.getLevel(), msg.getTime(), msg.getStackTraceElement());
                } else {
                    logStacktraceInternal(msg.getMessage(), msg.getLevel());
                }
            }
            pausedMessages.clear();
        }
    }

    public static void setEnabled(boolean enabled) {
        CCLogger.enabled = enabled;
    }

    public static boolean isEnabled() {
        return enabled;
    }

    public static File getLogFile() {
        return logFile;
    }

    public static void setGitHubCodePathURL(String url) {
        gitHubCodePathURL = url;
    }

    public static void setGitHubCodeClassPath(String classPath) {
        gitHubCodeClassPath = classPath;
    }

    public static void setLogFormat(String logFormat) {
        CCLogger.logFormat = logFormat;
    }

    public static void setLogLevel(CCLogLevel logLevel) {
        CCLogger.logLevel = logLevel;
    }

}

File: src/main/kotlin/org/capturecoop/cclogger/CCLogger.kt
package org.capturecoop.cclogger

import org.apache.commons.lang3.exception.ExceptionUtils
import org.apache.commons.text.StringEscapeUtils
import org.capturecoop.ccutils.utils.CCStringUtils
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import java.time.LocalDateTime
import javax.imageio.ImageIO

class CCLogger {
    companion object {
        val icon = ImageIO.read(CCDebugConsole::class.java.getResource("/org/capturecoop/cclogger/resources/console.png")) ?: throw FileNotFoundException("console.png not found!")
        val tabSpaces = 4
        var enabled = false
        var paused = false
            set(value) {
                field = value
                updatePaused()
            }
        var logFile: File? = null
            set(value) {
                value?.createNewFile()
                field = value
                refreshPreFileMessages()
            }
        private val MAX_LEVEL_LENGTH = CCLogLevel.WARNING.toString().length
        private val preFileMessages = ArrayList<String>() //These are printed but not written right away
        private val pausedMessages = ArrayList<CCLogMessage>() //These need to be constructed once ready
        var logFormat = "[%hour%:%minute%:%second%:%ms%] [%level%]%levelspace% [%filename%.%method%:%line%]: %message%"
        var htmlLog = ""
        var gitHubCodePathURL: String? = null //Example: https://github.com/CaptureCoop/SnipSniper/tree/<HASH HERE>/src/main/java/"
        var gitHubCodeClassPath: String? = null //Example: org.snipsniper -> If set only messages that contain this classpath get the @ link
        private var console: CCDebugConsole? = null
        private const val THIS_CLASSPATH = "org.capturecoop.cclogger.CCLogger"

        var filter = CCLogFilter.INFO

        fun info(message: Any?) = log(message, CCLogLevel.INFO)
        fun warn(message: Any?) = log(message, CCLogLevel.WARNING)
        fun error(message: Any?) = log(message, CCLogLevel.ERROR)
        fun debug(message: Any?) = log(message, CCLogLevel.DEBUG)

        fun log(message: Any?, level: CCLogLevel) {
            if(!enabled) return
            if(paused) {
                pausedMessages.add(CCLogMessage(level, message.toString(), LocalDateTime.now(), getStackTrace(), false))
                return
            }
            logInternal(message.toString(), level, LocalDateTime.now(), getStackTrace())
        }

        //The reason for this is that this way we can take index 3 of stack trace at all times
        private fun logInternal(message: String, level: CCLogLevel, time: LocalDateTime, currentStackTrace: StackTraceElement) {
            if(!enabled || filter == CCLogFilter.NONE) return
            when(filter) {
                CCLogFilter.ERRORS -> if(level.isNot(CCLogLevel.ERROR)) return
                CCLogFilter.WARNINGS -> if(level.isNot(CCLogLevel.ERROR, CCLogLevel.WARNING)) return
                CCLogFilter.INFO -> if(level.isNot(CCLogLevel.ERROR, CCLogLevel.WARNING, CCLogLevel.INFO)) return
                else -> {}
            }

            var msg = CCStringUtils.formatDateTimeString(logFormat, time)

            //Handle %level% and %levelspace%
            level.toString().also {
                var levelString = it
                if(levelString.length <= MAX_LEVEL_LENGTH) {
                    msg = msg.replace("%levelspace%", " ".repeat(MAX_LEVEL_LENGTH - levelString.length))
                } else {
                    levelString = levelString.substring(0, MAX_LEVEL_LENGTH)
                    msg = msg.replace("%levelspace%", "")
                }
                msg = msg.replace("%level%", levelString)
            }

            //Handle %filename% and %fileext%
            File(currentStackTrace.fileName ?: "null").also {
                msg = msg.replace("%filename%", it.nameWithoutExtension)
                msg = msg.replace("%fileext%", it.extension)
            }

            //Handle %method%, %line%, %message% and %newline%
            msg = msg.replace("%method%", currentStackTrace.methodName)
            msg = msg.replace("%line%", currentStackTrace.lineNumber.toString())
            msg = msg.replace("%message%", message)
            msg = msg.replace("%newline%", "\n")

            println(level.getAnsiColor() + msg + level.getAnsiReset())

            var finalMsg = StringEscapeUtils.escapeHtml4(msg).replace(" ", "&nbsp;")
            finalMsg = finalMsg.replace("%newline%", "<br>")
            if(gitHubCodePathURL != null && (gitHubCodeClassPath == null || gitHubCodeClassPath.isNullOrEmpty() || currentStackTrace.className.contains(gitHubCodeClassPath!!))) {
                //TODO: check that this works once the kotlin port of SnipSniper is complete
                val extension = File(currentStackTrace.fileName).extension
                val lineNr = currentStackTrace.lineNumber
                val link = gitHubCodePathURL + currentStackTrace.className.replace("\\.", "/") + ".$extension#L$lineNr"
                finalMsg = finalMsg.replace(":$lineNr]", ":$lineNr <a href='$link'>@</a>]")
            }
            htmlLog += "<p style='margin-top:0 white-space: nowrap'><font color='${level.getHTMLColor()}'>${finalMsg}</font></p>"
            console?.update()
            msg += "\n"
            writeToFile(msg)
        }

        fun logStacktrace(throwable: Exception, level: CCLogLevel) = logStacktrace(message = null, throwable = throwable, level = level)
        fun logStacktrace(throwable: Throwable, level: CCLogLevel) = logStacktrace(message = null, throwable = throwable, level = level)
        fun logStacktrace(message: String? = null, exception: Exception, level: CCLogLevel) = logStacktrace(message, throwable = exception, level)
        fun logStacktrace(message: String? = null, throwable: Throwable, level: CCLogLevel) = logStacktraceInternal(message ?: throwable.message ?: "No message", ExceptionUtils.getStackTrace(throwable), level)

        private fun logStacktraceInternal(message: String?, stackTraceString: String, level: CCLogLevel) {
            if(!enabled) return

            if(message != null) log(message, level)
            if(paused) {
                pausedMessages.add(CCLogMessage(level, stackTraceString, LocalDateTime.now(), null, true))
                return
            }

            println(level.getAnsiColor() + stackTraceString + level.getAnsiReset())
            writeToFile(stackTraceString)
            val htmlEscaped = StringEscapeUtils.escapeHtml4(stackTraceString).replace("\n", "<br>").replace("\t", "&nbsp;".repeat(tabSpaces))
            htmlLog += "<p style='margin-top:0 white-space: nowrap'><font color='${level.getHTMLColor()}'>${htmlEscaped}</font></p>"
            htmlLog += "<br>"
        }

        private fun updatePaused() {
            if(!paused) {
                pausedMessages.forEach { msg ->
                    if(!msg.isException) {
                        logInternal(msg.message, msg.level, msg.time, msg.stackTraceElement ?: throw Exception("Oops! CCLogger 169 issue, stackTraceElement is null!!"))
                    } else {
                        logStacktraceInternal(message = null, stackTraceString = msg.message, level = msg.level)
                    }
                }
                pausedMessages.clear()
            }
        }

        private fun refreshPreFileMessages() {
            preFileMessages.forEach { writeToFile(it) }
            preFileMessages.clear()
        }

        private fun writeToFile(message: String) {
            if(logFile != null) {
                try {
                    Files.write(Paths.get(logFile!!.absolutePath), message.toByteArray(), StandardOpenOption.APPEND)
                } catch (ioException: IOException) {
                    val path = logFile!!.absolutePath
                    logFile = null
                    log("Could not write to logfile at \"$path\". Disabling logFile & Printing to console as well just in case!", CCLogLevel.ERROR)
                    logStacktrace(ioException, CCLogLevel.ERROR)
                }
            } else {
                preFileMessages.add(message)
            }
        }

        private fun getStackTrace(): StackTraceElement {
            //We loop through the stacktrace till we get to whoever called the logger. We skip the first element as its java.lang.Thread
            val stackTrace = Thread.currentThread().stackTrace
            var startIndex = 1
            for(i in 1 until stackTrace.size) {
                if(stackTrace[startIndex].className.startsWith(THIS_CLASSPATH)) startIndex++
                else break
            }
            return stackTrace[startIndex]
        }

        fun enableDebugConsole(enabled: Boolean) {
            if(enabled) {
                if(console == null) console = CCDebugConsole()
                console?.update()
            } else {
                console?.isVisible = false
                console?.dispose()
                console = null
            }
        }
    }
}